java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -175,14 +174,45 @@ class Guard extends ExprParent {
   }
 }
 
+/**
+ * A `Guard` that tests an expression's type -- that is, an `instanceof T` or a
+ * `case T varname` pattern case.
+ */
+class TypeTestGuard extends Guard {
+  Expr testedExpr;
+  Type testedType;
+
+  TypeTestGuard() {
+    exists(InstanceOfExpr ioe | this = ioe |
+      testedExpr = ioe.getExpr() and
+      testedType = ioe.getCheckedType()
+    )
+    or
+    exists(PatternCase pc | this = pc |
+      pc.getSelectorExpr() = testedExpr and
+      testedType = pc.getPattern().getType()
+    )
+  }
+
+  /** Holds if this guard tests whether `e` has type `t`. */
+  predicate appliesTypeTest(Expr e, Type t) { e = testedExpr and t = testedType }
```

Comment:
This is not entirely correct for record patterns, as such patterns might correspond to an implicit conjunction with a sequence of nested type tests. As such, record patterns are always type tests on the true-edge, but only sometimes negative type tests on the false-edge, where the latter depends on the record pattern not containing nested type test that are capable of failing.

Reply:
Good point -- new test cases added and `appliesTypeTest` modified to account for that.